solr/solrj/src/java/org/apache/solr/common/cloud/Replica.java

```diff
@@ -167,11 +167,15 @@ public String getNodeName() {
     return nodeName;
   }
   
-  /** Returns the {@link State} of this replica. */
+  /** Returns the {@link State} of this replica. use {@link org.apache.solr.client.solrj.cloud.ShardStateProvider#getState(Replica)}  instead*/
+  @Deprecated
   public State getState() {
```

Comment:
To be removed in 9.0?

Reply:
This may not be removed in 9.0. This may impact almost 100% of SolrCloud users. We need to ensure that the new solution is more stable, performant and our users are on board with this. This may go away in 10.0

Reply:
Then I think this refactoring falls short of its potential - because instead of having a single uniform API we now end up with two APIs that claim to do the same, and the old one still prevents us from moving the state reporting out of the actual Replica object - ie. no matter how clever we are with the `ShardStateProvider` we still end up having to cache the potentially stale state inside `Replica`.

Reply:
> the old one still prevents us from moving the state reporting out of the actual Replica object..

We can only deprecate stuff . We could possibly remove it but that should be done later (may be 9.0)